pmd-apex/src/test/java/net/sourceforge/pmd/lang/apex/rule/security/ApexSharingViolationsNestedClassTest.java

```diff
@@ -0,0 +1,194 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.rule.security;
+
+import static org.junit.Assert.assertEquals;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import net.sourceforge.pmd.Report;
+import net.sourceforge.pmd.Rule;
+import net.sourceforge.pmd.RuleViolation;
+import net.sourceforge.pmd.lang.LanguageRegistry;
+import net.sourceforge.pmd.lang.apex.ApexLanguageModule;
+import net.sourceforge.pmd.testframework.RuleTst;
+
+/**
+ * <p>Sharing settings are not inherited by inner classes. Sharing settings need to be declared on the class that
+ * contains the Database method, DML, SOQL, or SOSL.</p>
+ *
+ * <p>This test runs against Apex code that has an Outer class and and Inner class. Different Apex code is generated
+ * based on the boolean permutations. Any classes that includes data access cod, but doesn't declare a sharing setting
+ * should trigger a violation.</p>
+ */
+@RunWith(Parameterized.class)
```

Comment:
The number of permutations that can occur lends itself to a parameterized test instead hand crafting them in XML.

Reply:
The downside is, that it's now not obvious anymore, what is being tested. You can't copy the sample code snippet into some other tool (like designer, or another parser/compiler, or easily create a separate source file).

Since we still have the "standard" rule tests, I'm ok with adding this test class.

Reply:
@adangel thanks for the review. Would you like to me to programatically generate the xml tests and only check in the xml. That would remove the need for this java class, and fix some of the concerns that you raised.

Reply:
@jbartolotta-sfdc I'd keep this class as it is for now. We can migrate that later to xml once we actually need it.